A tagging system for dashboards, charts and queries

betodealmeida opened the pull request:
This PR introduces a tagging system for Superset, allowing dashboards, charts and queries to be tagged. It also allows searching for a given tag, and will be the basis for a new landing page (see https://github.com/apache/incubator-superset/issues/5327).

# Adding tags

Here's an example showing adding tags to a chart (dashboards work the same way). Suggestions come from pre-existing tags, and are ordered by popularity:

![add_tag](https://user-images.githubusercontent.com/1534870/43427621-3564d936-940f-11e8-9912-6aba03280099.gif)

And here's how it looks like when a user is not an owner of the chart/dashboard:

<img width="419" alt="screen shot 2018-07-30 at 3 44 31 pm" src="https://user-images.githubusercontent.com/1534870/43427675-7ae8e326-940f-11e8-854b-293e8b90d377.png">

# Searching for tags

Clicking on a tag will take the user to a page searching for that tag. For example, clicking on the "test" tag:

<img width="1680" alt="screen shot 2018-07-30 at 3 45 50 pm" src="https://user-images.githubusercontent.com/1534870/43427705-a8b85e80-940f-11e8-8ed7-601fa52d439e.png">

For this, I added a new tab "Tags" to the welcome page, using the same layout as the "Favorites" tab. I changed the page so that the active tab can be specified by the URL anchor, as well as the search parameter. For example, the URL `/superset/welcome?q=test#tags` will show the "Tags" tab instead of the default "Dashboards" with the search form pre-populated with "test".

# Implicit tags

Dashboard, chart and (saved) queries have implicit tags related to their owners, types and favorites. For example, all objects owned by the admin have the tag `owner:1`. All charts have the tag `type:chart`. Objects favorited by the admin have the tag `favorited_by:1`.

These tags are automatically added by a migration script, and kept in sync through SQLAlchemy event listeners. They are currently not surfaced to the user, but can be searched for. For example, it's possible to search for `owner:1` in the welcome page to see all objects owned by the admin, or even search for `owner:{{ current_user_id() }}`.

# Queries

Currently queries cannot be tagged, because I wasn't sure were to add the component. Any suggestions, @vylc?

betodealmeida committed:
WIP

betodealmeida committed:
Merge branch 'master' of github.com:apache/incubator-superset into tagging_system

betodealmeida committed:
Missing update

betodealmeida committed:
Refactor tag component

betodealmeida committed:
Decouple tag component from app

betodealmeida committed:
Working on hooks

betodealmeida committed:
Ready for review

betodealmeida committed:
Minor adjustments

betodealmeida committed:
Small fixes

betodealmeida commented:
Fixing tests.

hughhhh commented on superset/views/tags.py:
```diff
@@ -0,0 +1,198 @@
+from .base import BaseSupersetView
+
+
+class ObjectTypeConverter(BaseConverter):
```
Can we move this to a `utils.py` file?

betodealmeida replied:
Just to be clear, this class is used for validating the path in the views defined in this file. For example:

```python
@expose('/tags/<object_type:object_type>/<int:object_id>/', methods=['GET'])
def get(self, object_type, object_id):
    ...
```

Using the validator, `/tags/chart/1/` is a valid path (since `chart` is in the `ObjectTypes` enum) but `/tags/foo/1` is invalid. Since the concept of `object_type` is only used in tag views, I think moving this to `utils.py` will not be very useful, since it's unlikely that it will be used in other places.

I'll add unit tests, good point.

hughhhh commented on superset/assets/src/explore/components/ExploreChartHeader.jsx:
```diff
@@ -32,9 +41,22 @@ const propTypes = {
+  constructor(props) {
+    super(props);
+
+    this.fetchTags = fetchTags.bind(this, 'chart', props.chart.id, false);
```
nit: you can use `autobind` here if you want 

https://github.com/andreypopp/autobind-decorator

betodealmeida replied:
But here I'm binding more than just `this`, I'm actually building a partial function that can be passed to `<ObjectTags />`. From what I understand `autobind` will only bind `this`, no?

hughhhh replied:
Ahh your right I thought there was way for additional params

vylc commented:
(1) Would suggest that the 'Altered' token goes before the tags, so as to reduce all the movement when you add tags
(2) What is the flow to search for charts/dashes with multiple tags? ie. if I wanted to search for 'test' and 'druid'. Seems like the Search box should be tokens instead of text?
(3) Re: queries. Not urgent, would rather get landing page shipped. We can put it in the dropdown menu for the query name tab
Looks good!

betodealmeida committed:
Fixing lint and addressing issues

betodealmeida committed:
Fix import order

betodealmeida committed:
Add polyfills

betodealmeida committed:
Fix polyfills

betodealmeida committed:
Fix pylint, javascript and mysql

betodealmeida committed:
Fix pylint and mysql

betodealmeida committed:
Fix javascript lint

betodealmeida committed:
Fix another unit test

betodealmeida committed:
Fix javascript tests

betodealmeida committed:
Fix for MySQL

betodealmeida commented:
Phew! So many fixes.

betodealmeida committed:
Tag search

betodealmeida commented:
@vylc I added better searching for tags:

![tag_search](https://user-images.githubusercontent.com/1534870/43550686-49e5c92a-9599-11e8-8e3c-ad9e3a0e1543.gif)

mistercrunch commented on superset/models/tags.py:
```diff
@@ -0,0 +1,233 @@
+            TaggedObject.object_id == target.obj_id,
+            Tag.type == TagTypes.favorited_by,
+            Tag.name == name,
+        )
```
I think you could just `query.delete()` at the next line here (as long as you change `session.query(TaggedObject.id)` -> `session.query(TaggedObject)`), no need for another query I think.

mistercrunch commented on superset/views/tags.py:
```diff
@@ -0,0 +1,197 @@
+
+    @expose('/tags/suggestions/', methods=['GET'])
+    def suggestions(self):
+        query = db.session.query(
```
nit: the way I like to break lines when doing method chaining in python is to wrap in parenthesis and break before the dot as in:
```python
query = (
    db.session.query(TaggedObject)
    .group_by(TaggedObject.tag_id)
    .order_by(func.count().desc())
    .all()
)
```

mistercrunch commented on superset/views/tags.py:
```diff
@@ -0,0 +1,197 @@
+        tags = json.dumps([
+            {'id': obj.tag.id, 'name': obj.tag.name} for obj in query])
+
+        return Response(tags, status=200, content_type='application/json')
```
I think you can use return `self.json_success(tags)` where tag is the object, not the json string.

optional side-mission: Looks like we may want to move https://github.com/apache/incubator-superset/blob/master/superset/views/core.py#L86 to `view/base.py` as a method of `BaseSupersetView`

mistercrunch replied:
If this new `BaseSupersetView.json_success` expects an object, it could be a way to standardize to simplejson.

betodealmeida replied:
Cool, I'll do that.

betodealmeida replied:
Actually, this is more complicated than I was expecting. The core view uses different encoding options depending on the visualization type (by calling `viz_obj.json_dumps`), so it requires a lot of refactoring.

Let's do that on a separate PR.

mistercrunch commented on superset/views/tags.py:
```diff
@@ -0,0 +1,197 @@
+        return Response(status=201)  # 201 CREATED
+
+    @expose('/tags/<object_type:object_type>/<int:object_id>/', methods=['DELETE'])
+    def delete(self, object_type, object_id):
```
Note that some of these endpoints we'd get for free by creating a FAB ModelView object, it may not be exactly what you need though, unclear. I'm ok with the more explicit/controlled endpoints.

betodealmeida replied:
👍

mistercrunch commented on superset/views/tags.py:
```diff
@@ -0,0 +1,197 @@
+                TaggedObject.object_id == SavedQuery.id,
+                TaggedObject.object_type == ObjectTypes.query,
+            ),
+        ).group_by(TaggedObject.object_id, TaggedObject.object_type)
```
Trying to understand what kind of SQL will come out of that query object... Is the group_by needed?

Also you may be able to parameterize the ObjectTypes as in:
```python
tagged_models = [Slice, Dashboard, Query]
for tm in tagged_models:
    query = query.outerjoin(tm, {...})
```
It would make it easier to add more tagged models in the future but may make the code more opaque.

betodealmeida replied:
The group by is needed because a user might search for tags "a" and "b", and if an object is tagged with both it would return duplicates.

mistercrunch reviewed (commented):
Wanted to release this round of review on the Python side, I'll do another round on the JS side, coming soon.

mistercrunch commented on superset/assets/src/dashboard/components/Header.jsx:
```diff
@@ -6,12 +6,21 @@ import HeaderActionsDropdown from './HeaderActionsDropdown';
 import { t } from '../../locales';
 import { UNDO_LIMIT, SAVE_TYPE_OVERWRITE } from '../util/constants';
 
+const CSRF_TOKEN = (document.getElementById('csrf_token') || {}).value;
```
I'm not sure about doing this, and less sure about doing this in module scope. If it's not in the redux store already, we could make sure it makes its way through here:
https://github.com/apache/incubator-superset/blob/4bf69a7260f1d4c8fbfe9d9c0300cae39ca3bf81/superset/views/base.py#L93

betodealmeida replied:
Yeah, I was unsure about this but it's being done in other parts of the code. Let me see if I can fix it.

betodealmeida replied:
@mistercrunch, I'm not sure how to avoid doing this. It seems to be in line with [the docs for CSRF in Flask](http://flask-wtf.readthedocs.io/en/stable/csrf.html#javascript-requests): the token is only present in the DOM if the user is authenticated (since it's added by the template), so my understanding is that it should be fine. And the token is already there in the DOM at this point, assigning it to `CSRF_TOKEN` here is just a shortcut.

I looked at other AJAX requests in Superset. They're using jQuery after calling `initJQueryAjax`, which also loads the token into a `$.settings` so it can be reused in all subsequent requests.

mistercrunch replied:
Gotcha, now I'd rather not having it in module scope for the implications it has that the browser exist, which could be an issue while importing during unit tests or for code introspection. Simply importing the file in node might spit out something about `document` not existing.

mistercrunch commented on superset/assets/src/welcome/Tags.jsx:
```diff
@@ -0,0 +1,151 @@
+  });
+}
+
+export class Tags extends React.PureComponent {
```
Maybe `TagsTables` or `TagsTable` as a component name?

betodealmeida replied:
Definitely, will change.

mistercrunch commented on superset/assets/src/welcome/Tags.jsx:
```diff
@@ -0,0 +1,151 @@
+      this.setState({ objects });
+    });
+  }
+  renderTable(type) {
```
How about not rendering the table if it is empty?

betodealmeida replied:
I'm fine with either way, but I tend slightly towards showing an empty table because it shows that there's no data. Otherwise the user might think that the data failed to load.

mistercrunch reviewed (commented):
Haven't done a full pass on the js code, but pushing this set of comment as I head out.

betodealmeida commented on superset/models/tags.py:
```diff
@@ -0,0 +1,233 @@
+Session = sessionmaker(autoflush=False)
+
+
+class TagTypes(enum.Enum):
```
Yeah, I think even in 2.7.3. There's also a backport `enum34` which we could add as an optional dependency, something like:

```python
# setup.py
if sys.version_info < (2, 7, 3):
    deps.append('enum34')
```

And I'm +1 on moving towards newer versions. I think when developing libraries it's important to be more conservative, but for applications it's fine.

betodealmeida commented on superset/models/tags.py:
```diff
@@ -0,0 +1,233 @@
+    object_type = 'query'
+
+    @classmethod
+    def get_owners_ids(cls, target):
```
👍

betodealmeida commented on superset/assets/src/components/ObjectTags.jsx:
```diff
@@ -0,0 +1,122 @@
+      addTag,
+      editable,
+      onChange,
+      ...rest } = this.props;
```
Yeah, I couldn't pass `...this.props` to the children because of `onChange`, so I decided to remove all the props the component actually uses and pass down the rest.

betodealmeida commented on superset/assets/src/welcome/Tags.jsx:
```diff
@@ -0,0 +1,151 @@
+  search: PropTypes.string,
+};
+
+export function fetchTags(objectType, objectId, includeTypes, callback) {
```
Good point, will fix.

betodealmeida committed:
Address comments

betodealmeida committed:
Fix conflict

betodealmeida committed:
Merge branch 'master' into tagging_system

betodealmeida committed:
Fix liny

betodealmeida committed:
Merge branch 'tagging_system' of github.com:lyft/incubator-superset into tagging_system

Commit:
Update constants.js

Make prettier happy.

betodealmeida committed:
Fix conflict

betodealmeida committed:
Merge branch 'tagging_system' of github.com:lyft/incubator-superset into tagging_system

mistercrunch commented on superset/assets/package.json:
```diff
@@ -80,6 +80,8 @@
     "mustache": "^2.2.1",
     "nvd3": "1.8.6",
     "parse-iso-duration": "^1.0.0",
+    "po2json": "^0.4.5",
```
does that belong here?

mistercrunch commented on superset/assets/package.json:
```diff
@@ -80,6 +80,8 @@
     "mustache": "^2.2.1",
     "nvd3": "1.8.6",
     "parse-iso-duration": "^1.0.0",
+    "po2json": "^0.4.5",
```
does that belong here? should be in dev if anywhere I think

betodealmeida replied:
That was added by someone else and came with a merge conflict in `package.json`, let me remove it.

mistercrunch commented on superset/assets/package.json:
```diff
@@ -80,6 +80,8 @@
     "nvd3": "1.8.6",
     "parse-iso-duration": "^1.0.0",
+    "po2json": "^0.4.5",
+    "promise-polyfill": "8.0.0",
```
I think for this to take effect it needs to be imported where you use promises. Though we do use promises already without having the polyfill I think...

betodealmeida replied:
The `fetch` polyfill uses this.

betodealmeida committed:
Merge heads, remove dep

betodealmeida committed:
Move CSRF token to function

mistercrunch commented on superset/assets/package.json:
```diff
@@ -121,7 +123,9 @@
-    "viewport-mercator-project": "^5.0.0"
+    "url-polyfill": "^1.0.13",
+    "viewport-mercator-project": "^5.0.0",
+    "whatwg-fetch": "^2.0.4"
```
Would ❤️ approval from @williaster and/or @graceguo-supercat on this import as a preferred method to eventually replace jQuery in the codebase

mistercrunch replied:
What I meant was not jquery vs fetch, but npm's `fetch` package vs `whatwg-fetch` which are different things. I have no opposition with either but should make a conscious decision.

williaster commented on superset/assets/src/welcome/App.jsx:
```diff
@@ -53,7 +96,7 @@ export default class App extends React.PureComponent {
             </Panel>
           </Tab>
-          <Tab eventKey={3} title={t('Favorites')}>
+          <Tab eventKey={'favorites'} title={t('Favorites')}>
```
don't need `{}`

betodealmeida replied:
👍

williaster reviewed (changes requested):
I feel like there are some new unneeded dependencies added here, and left a few comments.

williaster commented:
aesthetically can you also make the text of the tags centered vertically? there is more padding on the bottom right now.

![image](https://user-images.githubusercontent.com/4496521/43733103-5ee84c8a-9968-11e8-888f-99e8b0e3fb18.png)

betodealmeida commented on superset/assets/package.json:
```diff
@@ -80,6 +80,7 @@
     "mustache": "^2.2.1",
     "nvd3": "1.8.6",
     "parse-iso-duration": "^1.0.0",
+    "promise-polyfill": "8.0.0",
```
We don't need this, any promise polyfill should work.

betodealmeida commented on superset/assets/package.json:
```diff
@@ -121,7 +123,9 @@
     "underscore": "^1.8.3",
     "urijs": "^1.18.10",
-    "viewport-mercator-project": "^5.0.0"
+    "url-polyfill": "^1.0.13",
```
This is a WHATWG spec for constructing and parsing URLs (https://developer.mozilla.org/en-US/docs/Web/API/URL), not routing. Does `react-router` offer something similar?

betodealmeida committed:
Remove dep

betodealmeida commented on superset/assets/src/components/ObjectTags.jsx:
```diff
@@ -0,0 +1,122 @@
+  }
+
+  handleAddition(tag) {
+    const tags = [].concat(this.state.tags, tag);
```
⚡️

betodealmeida commented on superset/assets/src/modules/utils.js:
```diff
@@ -231,10 +231,17 @@ export function customizeToolTip(chart, xAxisFormatter, yAxisFormatters) {
   });
 }
 
+export function getCSRFToken() {
```
😄

betodealmeida commented on superset/assets/src/components/ObjectTags.jsx:
```diff
@@ -0,0 +1,122 @@
+        {...rest}
+        tags={this.state.tags}
+        suggestions={this.state.suggestions}
+        handleDelete={this.handleDelete.bind(this)}
```
Ah, good point, I forgot they will get recreated every time this way. Will move to the constructor.

betodealmeida commented on superset/assets/src/components/ObjectTags.jsx:
```diff
@@ -0,0 +1,122 @@
+        tags={this.state.tags}
+        suggestions={this.state.suggestions}
+        handleDelete={this.handleDelete.bind(this)}
+        handleAddition={this.handleAddition.bind(this)}
```
⚡️

betodealmeida commented on superset/assets/src/welcome/App.jsx:
```diff
@@ -35,16 +78,16 @@ export default class App extends React.PureComponent {
             </Panel>
           </Tab>
-          <Tab eventKey={2} title={t('Recently Viewed')}>
+          <Tab eventKey={'recent'} title={t('Recently Viewed')}>
```
+1

betodealmeida commented on superset/assets/src/welcome/App.jsx:
```diff
@@ -13,19 +18,57 @@ const propTypes = {
+  }
+  componentDidMount() {
+    fetchSuggestions(false, (suggestions) => {
+      const tagSuggestions = STANDARD_TAGS.concat(
```
⚡️

elibrumbaugh commented:
Hey @betodealmeida thanks for sharing this proposal! I've written up some feedback.
![superset_tag_feedback](https://user-images.githubusercontent.com/2042462/43749639-c7e13ce2-99a9-11e8-8008-0b5f1f05bbfd.png)

betodealmeida commented on superset/assets/src/tags.js:
```diff
@@ -0,0 +1,78 @@
+import 'whatwg-fetch';
+import { getCSRFToken } from './modules/utils';
+
+export function fetchTags(objectType, objectId, includeTypes, callback) {
```
Will add it.

betodealmeida commented on superset/assets/src/tags.js:
```diff
@@ -0,0 +1,78 @@
+  });
+}
+
+export function addTag(objectType, objectId, includeTypes, tag, callback, error) {
```
Will fix.

betodealmeida commented on superset/assets/src/welcome/TagsTable.jsx:
```diff
@@ -0,0 +1,89 @@
+    }
+  }
+  fetchResults(search) {
+    fetchObjects(search, null, (data) => {
```
⚡️

betodealmeida commented on superset/assets/src/welcome/TagsTable.jsx:
```diff
@@ -0,0 +1,89 @@
+import { t } from '../locales';
+
+const propTypes = {
+  search: PropTypes.string,
```
:zap:

betodealmeida committed:
Working on comments

betodealmeida committed:
Fix merge

betodealmeida committed:
Small fixes

betodealmeida committed:
Fix CSS

betodealmeida committed:
Fix unit test

betodealmeida commented:
@williaster, I addressed all the comments except for using `react-router` — it seems to me using `react-router` would require a lot of changes to the welcome app, making this PR even more complex. IMHO it would be better to start migrating to `react-router` in a separate PR. What do you think?

Also, I remembered we're already using `URL` in [a couple places](https://github.com/apache/incubator-superset/search?q=%22new+URL%28%22&unscoped_q=%22new+URL%28%22) without including the polyfill (mea culpa), so it might better to add the polyfill now and migrate to `react-router` later.

betodealmeida commented:
Yeah, I agree reducing the number of deps is better. Let me rewrite this to use `urijs`.

betodealmeida committed:
Fix merge conflict

betodealmeida committed:
Use uri.js, fix SQL Lab redirect

betodealmeida commented:
@williaster, I've replaced the URL polyfill with `urijs` (which actually has a nicer interface, resulting in smaller code!).

betodealmeida committed:
Migration script heads

betodealmeida commented:
@elibrumbaugh, thanks for the feedback!

> I can’t tell based on the .gif what visual affordances in the UI tell the dashboard owner this is where they add a tag.
> Dashboards can have really long names. The example currently shown is with a best case dashboard name and not faced with the space issue.

Good point, right now this is far from optimal:

![width](https://user-images.githubusercontent.com/1534870/43864722-5600508c-9b15-11e8-966c-91a024ea2215.gif)

I'll see if I can make the component have a fix width and have the tags wrap. Thoughts on this?

> The tag visuals clash with the visuals for altered. Side note the white text on yellow tags isn't accessible. The blue isn't either.

The colors come from the Bootstrap theme. I created https://github.com/apache/incubator-superset/issues/5590 to address the problem separately.

> You could place the tags in the dropdown menu as well. Quickly accessible but not reducing data ink.
> Why not allow the dash owner to add these on the edit dashboard metadata page?
> Alt. why not allow the dash owner to add these in a modal that is linked from the dropdown with other similar actions?

From a UX point of view we think they should not be hidden, neither to read nor to write. Let's talk about this Friday during our meeting? I'm ok with changing the way tags are created — it should be easy to change that while keeping the data model and the REST endpoints from this PR.

> I like the tag link to a search experience that captures all content with that tag.

😄

vylc commented:
@elibrumbaugh can you think of a good way to get back to that overall landing page w/out having to click on a tag? not completely intuitive clicking on a tab will take you back to that page

betodealmeida committed:
Default tag search to owned content

betodealmeida committed:
Simplify tag options

betodealmeida commented:
@williaster, are you ok with moving forward and merging this PR, and then changing the way the tag component is display in dashboards?

williaster commented:
@betodealmeida in the meetup last week I thought we discussed possibly splitting this into two parts: one for backend and one for frontend?

So I'd propose either
1) just merge backend here, do explore/dashboard frontend in another PR or
2) do both here if you don't want to split it

I think we should not merge PRs that affect UI/UX if the UI/UX is not where we want it to be.

betodealmeida committed:
Rebase and fix conflict

betodealmeida committed:
Remove ObjectTags input

betodealmeida committed:
Merge heads in alembic

betodealmeida committed:
Fix unit test

betodealmeida commented:
@williaster, I removed the tag input from this PR, leaving only the backend.

williaster commented on superset/assets/src/components/ObjectTags.jsx:
```diff
@@ -0,0 +1,126 @@
+    return (
+      <div className="react-tags-rw">
+        {this.state.tags.map(tag => (
+          <Label bsStyle="info">
```
this needs a `key` prop as [explained here](https://reactjs.org/docs/lists-and-keys.html).

betodealmeida committed:
Merge

betodealmeida commented:
Moved the backend to https://github.com/apache/incubator-superset/pull/6823.

1AB9502 commented:
@betodealmeida ,

this looks so good, is there a link to the frontend files?

betodealmeida commented:
@1AB9502 I'll post the PR for the frontend today for review.

betodealmeida commented:
@1AB9502 still need some fixes, but here it is: https://github.com/apache/incubator-superset/pull/7418

1AB9502 commented:
@betodealmeida 

Huge thanks!

junlincc commented:
@mihir174 @michael-s-molina @steejay 

Hey both, not sure if you were aware of this PR. Tagging was once bought up with both BE and FE solutions, it was closed because the proposed UI/UX was not ideal. To bring tagging(BE) back to life, we will have to rewrite a big trunk of work basically, but still, a shortcut to solve the content organization problem. 
We were wondering if we can use tagging for grouping, searching, hiding even temporarily deleting objects.
